### PR TITLE
Gateway actors for serving multiple federations

### PR DESCRIPTION
**Pull request by okjodom**
**Merge after #868**

Implement a simplified gataeway actor pattern to support running execution of gateway functions for multiple federations. A gateway actor is a 1:1 representation of a Federation in the context of the gateway. The actor:
- encapsulates a federation client
- exposes a logical api on the federation client, with the following functions possible:
  - `sync_wallet` - Fetch coins this gateway owns in the federation and settle the wallet state. Future: we should find a way of automating this sync operation
  - `buy_preimage_offer` - Buy a lightning preimage listed for sale inside the federation served by this actor
  - `await_preimage_decryption` - Wait for preimage decryption by the federation served by this actor
  - `pay_invoice` - Pay an invoice on behalf of a user in the federation served by this actor
  - `buy_preimage_internal` - Buy a preimage when completing an internal payment in the federation served by this actor
  - `buy_preimage_external` - Buy a preimage when completing an external payment for the federation
  - `await_outgoing_contract_claimed` - Wait for outgoing contract confirmation by the federation served by this actor
  - `get_deposit_address` - Get an address for depositing funds (peg-in) to the federation served by this actor
  - `deposit` - Deposit funds (peg-in) to the federation served by this actor
  - `withdraw` - Withdraw funds (peg-out) from the federation served by this actor
  - `get_balance` - Get the current e-cash balance this gateway has in the federation served by this actor

Externally, the gateway exposes an api `register_federation`. This creates a new actor for servicing the federation.

### Beyond this PR, we could:
- [ ] #664
  - [x] - implement an external ( webserver & cli ) api to support federation registration
  - [ ] - add federation identifier information in the routing hint. Add an extra hop for this (?)
- [ ] Run actor actions in isolated threads for perf
- [ ] Implement **self payments** where a single gateway facilitates payments across two federations it serves
- [ ] Test, test, test!

### CODE REVIEW

File: ln-gateway/src/actor.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+        GatewayActorCtrl {
+            is_actor_shutting_down: Arc::clone(&self.is_actor_shutting_down),
+        }
+    }
+}
+
+pub struct GatewayActorCtrl {
+    is_actor_shutting_down: Arc<AtomicBool>,
+}
```
Comment (justinmoon) [inline ln-gateway/src/actor.rs:114]: What is the difference between this and the `is_actor_shutting_down` inside `GatewayActor`?

File: ln-gateway/src/actor.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+impl GatewayActorCtrl {
+    pub fn stop(&self) {
+        self.is_actor_shutting_down.store(true, Ordering::SeqCst);
+    }
+}
```
Comment (justinmoon) [inline ln-gateway/src/actor.rs:120]: Why make a new struct for this instead of just putting `stop()` on `GatewayActor`?

Comment (okjodom) [reply ln-gateway/src/actor.rs:120]: I added this struct before I figured I needed to extract a `GatewayCore` thing. A simple `stop()` is nice, so I will adopt that when I iterate

File: fedimintd/src/bin/configgen.rs [hunk 3]
```diff
@@ -47,7 +47,7 @@ enum Command {
         denominations: Vec<Amount>,
 
         /// Federation name
-        #[arg(long = "federation-name", default_value = "Hal's trusty mint")]
+        #[arg(long = "federation-name", default_value = "Hals_trusty_mint")]
```
Comment (okjodom) [inline fedimintd/src/bin/configgen.rs:50]: Woops, **"Hal's trusty mint"** got a new name! One that's more CLI friendly! :)

File: gateway/ln-gateway/src/lib.rs [hunk 4]
```diff
@@ -188,6 +188,14 @@ impl LnGateway {
         Ok(actor)
     }
 
+    fn select_actor(&self, federation_id: FederationId) -> Result<Arc<GatewayActor>> {
+        // TODO: Select federation client based on federation id
+        self.actors
+            .get(&federation_id)
+            .cloned()
+            .ok_or(LnGatewayError::UnknownFederation)
+    }
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:196]: Isn't it doing exactly this here?

File: gateway/ln-gateway/src/lib.rs [hunk 5]
```diff
@@ -295,12 +295,16 @@ impl LnGateway {
...
 
-    async fn handle_deposit_msg(&self, deposit: DepositPayload) -> Result<TransactionId> {
-        let rng = rand::rngs::OsRng;
-        self.federation_client
-            .peg_in(deposit.txout_proof, deposit.transaction, rng)
+    async fn handle_deposit_msg(&self, payload: DepositPayload) -> Result<TransactionId> {
+        let DepositPayload {
+            txout_proof,
+            transaction,
+            federation_id,
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:302]: Having seen the same pattern a few times now, could we extract the `federation_id` into another struct that contains the ID and the actual request?

**Review by elsirion — APPROVED** [review]
Looks good

File: gateway/ln-gateway/src/lib.rs (hunk 4 above)
Comment (okjodom) [reply gateway/ln-gateway/src/lib.rs:196]: Oh, I missed cleaning up the TODO comment!

**Review by justinmoon — COMMENTED** [review]
Looks great. A few small nits.

File: gateway/ln-gateway/src/lib.rs [hunk 6]
```diff
@@ -334,38 +283,39 @@ impl LnGateway {
...
+        } = payload;
+
+        self.select_actor(federation_id)?
+            .withdraw(amount, address)
             .await
-            .map_err(LnGatewayError::ClientError)
-            .map(|out_point| out_point.txid)
     }
 
     pub async fn run(&mut self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (okjodom) [reply gateway/ln-gateway/src/lib.rs:321]: In a follow up PR? Creating this loop for fetching coins would come nicely with optimization to run actors actions in distinct threads...

**Review by justinmoon — APPROVED** [review]

**Pull request merged**